crypto/s2n_libcrypto.c

```diff
@@ -134,6 +138,21 @@ bool s2n_libcrypto_is_boringssl()
 #endif
 }
 
+S2N_RESULT s2n_libcrypto_init(void)
+{
+#if S2N_OPENSSL_VERSION_AT_LEAST(3, 0, 0)
+    RESULT_ENSURE(OSSL_PROVIDER_load(NULL, "default") != NULL, S2N_ERR_OSSL_LOAD_PROVIDER);
+    #ifdef S2N_LIBCRYPTO_SUPPORTS_EVP_RC4
+    /* needed to support RC4 algorithm
+     * https://www.openssl.org/docs/man3.0/man7/OSSL_PROVIDER-legacy.html
+     */
+    RESULT_ENSURE(OSSL_PROVIDER_load(NULL, "legacy") != NULL, S2N_ERR_OSSL_LOAD_PROVIDER);
+    #endif
```

Comment:
Do you need to call OSSL_PROVIDER_unload from s2n_cleanup? Are we somehow leaking resources if we don't?

> OSSL_PROVIDER_unload() unloads the given provider. For a provider added with OSSL_PROVIDER_add_builtin(), this simply runs its teardown function.

Reply:
--- short version
no i dont think so. `unload` based on the documentation is meant as a tool to stop using a functionality rather than a cleanup mechanism for providers which come packaged with openssl.

--- long version

By default openssl3 comes with a few providers available for loading: [default](https://www.openssl.org/docs/man3.0/man7/OSSL_PROVIDER-default.html), [base](https://www.openssl.org/docs/man3.0/man7/OSSL_PROVIDER-base.html), [legacy](https://www.openssl.org/docs/man3.0/man7/OSSL_PROVIDER-legacy.html), etc. Additionally the `default` provider is loaded automatically if no other provider is loaded.

`OSSL_PROVIDER_add_builtin`: `add a built in provider to OSSL_PROVIDER store in the given library context`. which means I can add a custom providers which could have special guarantees. Imagine "my_special_fips_provider".

`OSSL_PROVIDER_load`: `loads and initializes a provider. This may simply initialize a provider that was previously added with OSSL_PROVIDER_add_builtin()`. which mean even though the provider is added to the store, it must still be loaded to be used within a library context.

`OSSL_PROVIDER_unload(): `unloads the given provider. For a provider added with OSSL_PROVIDER_add_builtin(), this simply runs its teardown function`; which means unload the provider (dont use it). This can be used to unload the provider dynamically and make sure certain functionality(algorithms) are not available. Imagine if I want to load and unload `legacy` provider dynamically. The teardown function seems to be run if the custom provider was added via the `add_builtin` function. Since we only use the `default` and `legacy` providers, I dont think we need to do an unload, which is also not necessary based on the documentation.

Reply:
turns out we do need to cleanup. running asan tests didnt catch it but valgrind did: https://github.com/aws/s2n-tls/pull/3481